Definition of Insitu<Plugin> class and use in MySimulation class

anshumang opened the pull request:
This branch's purpose is to serialize PIConGPU's particle data for delivery to a separate analysis application (not running in PIConGPU's address space).

- InsituBinEnergyParticles class cloned from BinEnergyParticles class
- load, unload and notify called from MySimulation class's load, unload and runOneStep respectively

anshumang commented:
There is already dumpOneStep which can write species data using ADIOSWriter. So runOneStep might not be the appropriate place to call InsituBinEnergyParticles::notify. I want to measure if insitu is possible even when dumping in every step. I have it here now to avoid any additional overhead of going through dumpOneStep in each step.

ax3l commented:
hi, welcome and thx for your first pull request.

please register a plugin not in `MySimulation.hpp` but in `src/picongpu/include/plugins/PluginController.hpp`.

I am not really seeing a implementation in the new plugin yet - is that on purpose?

anshumang commented:
Moved registration to PluginController. Cleaned up the InsituBinEnergyParticles class further. Yes, the implementation is missing because most of it is the same as ADIOSWriter. The reason I want to create a new plugin is because the Landrush/Goldrush library is not part of ADIOS yet. So this only needs the particle data serialization code from ADIOSwriter, none of the ADIOS specific stuff.

@ax3l  Can you please check my commit author information now? I updated my local git config.

anshumang commented:
ADIOSWriter launches kernelCountParticles and copySpecies to count and copy particle data. Can I not copy all the data over to the host memory and perform the strided access there?

ax3l commented on src/picongpu/include/simulationControl/MySimulation.hpp:
```diff
@@ -244,6 +244,7 @@ class MySimulation : public SimulationHelper<simDim>
         __delete(pushBGField);
         __delete(currentBGField);
         __delete(cellDescription);
+        
```
pls try to avoid changes in this file for now :)

ax3l commented:
All right, I think you can continue to work on `InsituBinEnergyParticles.hpp` and implement your ideas. If you want to have a look on the current adios implementation, please refer to #679 which is changing a lot currently and will be merged soon.

ax3l commented:
@anshumang I'll close the pull request for now. If you need new feedback on a updated version just open a new one :)

anshumang commented:
Sure @ax3l !
